rllib/models/tf/tf_action_dist.py

```diff
@@ -64,32 +64,25 @@ def deterministic_sample(self) -> TensorType:
 
     @override(ActionDistribution)
     def logp(self, x: TensorType) -> TensorType:
-        return -tf.nn.sparse_softmax_cross_entropy_with_logits(
-            logits=self.inputs, labels=tf.cast(x, tf.int32)
-        )
+        logp = tf.math.log(tf.nn.softmax(self.inputs))
+        return tf.gather(logp, x, batch_dims=1)
 
     @override(ActionDistribution)
     def entropy(self) -> TensorType:
-        a0 = self.inputs - tf.reduce_max(self.inputs, axis=1, keepdims=True)
-        ea0 = tf.exp(a0)
-        z0 = tf.reduce_sum(ea0, axis=1, keepdims=True)
-        p0 = ea0 / z0
-        return tf.reduce_sum(p0 * (tf.math.log(z0) - a0), axis=1)
+        probs = tf.nn.softmax(self.inputs)
+        probs = tf.clip_by_value(probs, tf.keras.backend.epsilon(), 1.0)
+        return tf.reduce_sum(probs * tf.math.log(probs), axis=1)
 
     @override(ActionDistribution)
     def kl(self, other: ActionDistribution) -> TensorType:
-        a0 = self.inputs - tf.reduce_max(self.inputs, axis=1, keepdims=True)
-        a1 = other.inputs - tf.reduce_max(other.inputs, axis=1, keepdims=True)
-        ea0 = tf.exp(a0)
-        ea1 = tf.exp(a1)
-        z0 = tf.reduce_sum(ea0, axis=1, keepdims=True)
-        z1 = tf.reduce_sum(ea1, axis=1, keepdims=True)
-        p0 = ea0 / z0
-        return tf.reduce_sum(p0 * (a0 - tf.math.log(z0) - a1 + tf.math.log(z1)), axis=1)
+        probs = tf.nn.softmax(self.inputs)
+        probsOther = tf.nn.softmax(other.inputs)
+        return tf.keras.losses.kl_divergence(probs, probsOther)
 
     @override(TFActionDistribution)
     def _build_sample_op(self) -> TensorType:
-        return tf.squeeze(tf.random.categorical(self.inputs, 1), axis=1)
+        logp = tf.math.log(tf.nn.softmax(self.inputs))
```

Comment:
Same here, why don't we clip by value with tf.keras.backend.epsilon() here, too?

Reply:
I approve using logprobs here though as it common practice and adds stability.